src/osd/OSDMap.cc

```diff
@@ -1639,18 +1639,17 @@ void OSDMap::dedup(const OSDMap *o, OSDMap *n)
 }
 
 void OSDMap::clean_temps(CephContext *cct,
-			 const OSDMap& osdmap, Incremental *pending_inc)
+			 const OSDMap& oldmap,
+			 const OSDMap& tmpmap,
+			 Incremental *pending_inc)
 {
   ldout(cct, 10) << __func__ << dendl;
-  OSDMap tmpmap;
-  tmpmap.deepish_copy_from(osdmap);
-  tmpmap.apply_incremental(*pending_inc);
 
   for (auto pg : *tmpmap.pg_temp) {
     // if pool does not exist, remove any existing pg_temps associated with
     // it.  we don't care about pg_temps on the pending_inc either; if there
     // are new_pg_temp entries on the pending, clear them out just as well.
-    if (!osdmap.have_pg_pool(pg.first.pool())) {
+    if (!tmpmap.have_pg_pool(pg.first.pool())) {
```

Comment:
should this be `if (!oldmap.have_pg_pool(pg.first.pool()))`?

Reply:
I think this is an improvement over the previous code because it will remove new pg_temp entries if the pool is simultaneously removed.  The old code would end up with one osdmap epoch with the useless entries and remove them in the next epoch.